src/Microsoft.AspNetCore.Blazor.Server/Circuits/BlazorHub.cs

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Blazor.Services;
+using Microsoft.AspNetCore.SignalR;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    internal class BlazorHub : Hub
+    {
+        private static readonly object CircuitKey = "circuit";
+
+        private readonly CircuitFactory _circuitFactory;
+
+        public BlazorHub(CircuitFactory circuitFactory)
+        {
+            _circuitFactory = circuitFactory ?? throw new ArgumentNullException(nameof(circuitFactory));
+        }
+
+        public CircuitHost CircuitHost
+        {
+            get => (CircuitHost)Context.Items[CircuitKey];
+            set => Context.Items[CircuitKey] = value;
+        }
+
+        public override Task OnDisconnectedAsync(Exception exception)
+        {
+            CircuitHost.Dispose();
+            return base.OnDisconnectedAsync(exception);
+        }
+
+        public async Task StartCircuit(string uriAbsolute, string baseUriAbsolute)
+        {
+            var circuitHost = _circuitFactory.CreateCircuit(Context.GetHttpContext(), Clients.Caller);
+
+            var uriHelper = circuitHost.Services.GetRequiredService<IUriHelper>() as RemoteUriHelper;
+            if (uriHelper != null)
```

Comment:
As it stands, it looks like there will always be a `IUriHelper` because it's set up in `AddBlazorCircuits`.

I can't think of any important reason why we should make it optional. But if we are having it mandatory, then:

 * Probably don't need to check for nullness here
 * We can probably do away with the `enableNavigationInterception` call from .NET to JS, and have `Boot.WebAssembly.ts` call it directly on startup

Reply:
this check isn't for optionality, it's in case the UriHelper was replaced. I'm also ok with saying that's not supported and just making it a hard cast.

On `enableNavigationInterception` - I totally agree, that would simplify things.